ext/drreg/drreg.c

```diff
@@ -437,6 +669,73 @@ drreg_event_bb_insert_late(void *drcontext, void *tag, instrlist_t *bb, instr_t
     }
 
     /* Before each app read, or at end of bb, restore spilled registers to app values: */
+    for (reg = DR_REG_APPLICABLE_START_SIMD; reg <= DR_REG_APPLICABLE_STOP_SIMD; reg++) {
+        restored_for_simd_read[SIMD_IDX(reg)] = false;
+        if (!pt->simd_reg[SIMD_IDX(reg)].native) {
+            if (drmgr_is_last_instr(drcontext, inst) ||
+                instr_reads_from_reg(inst, reg, DR_QUERY_INCLUDE_ALL) ||
+                /* i#1954: for complex bbs we must restore before the next app instr */
+                (!pt->simd_reg[SIMD_IDX(reg)].in_use &&
+                 ((pt->bb_has_internal_flow &&
+                   !TEST(DRREG_IGNORE_CONTROL_FLOW, pt->bb_props)) ||
+                  TEST(DRREG_CONTAINS_SPANNING_CONTROL_FLOW, pt->bb_props)))) {
+                if (!pt->simd_reg[SIMD_IDX(reg)].in_use) {
+                    res = drreg_restore_reg_now(drcontext, bb, inst, pt, reg);
+                    if (res != DRREG_SUCCESS)
+                        drreg_report_error(res, "lazy restore failed");
+                    ASSERT(pt->simd_pending_unreserved > 0, "should not go negative");
+                    pt->simd_pending_unreserved--;
+                } else {
+                    if (reg_is_vector_simd(reg)) {
```

Comment:
Can we be here and it's not vector simd?

Reply:
Done - thanks for pointing this out!